Race conditions when updating PlanningScene: fixup #716

rhaschke opened the pull request:
This continues #716 and #724 which were both merged already. I repeat here my [comments](https://github.com/ros-planning/moveit_ros/pull/724#issuecomment-235284344) to #724:

@v4hn Realizing the simplification we agreed on in the phone call turned out to be rather difficult:
1. simplification of `stateUpdateTimerCallback()` (removing the additional timestamp check) caused dead locks. Hence, I reverted these changes.
2. I successfully moved `validatePlan()` from `MoveGroupInterface` (client side) to `MoveGroup`'s `ExecutionServiceCapability` (server side) - as agreed.
3. `Simplification of syncSceneUpdates()`: We definitely need both while loops and the timeout. When the robot doesn't move at all, we will never receive a scene update. The first loop is required to check for a recent robot state update directly in CSM, while the second loop (in case there is no CSM) waits - with a timeout - for a direct scene update. There is no way to omit this timeout. The only chance would be to trigger all input channels of PSM to send updates - which is not possible. However, in `move_group`, this doesn't pose a problem, because there is a CSM instantiated.
4. Removing `syncSceneUpdates()` after trajectory execution makes sense. Doing so, led to failure of the [test](https://gist.github.com/JeremyZoss/63dfdf24a97c5e6e7d65) suggested in #442. Now I fixed it. `CSM::waitForCurrentState()` didn't do what I expected from the name...

Some more thoughts:
- I suggest some more API changes: `syncSceneUpdates()` should be renamed to `waitForCurrentRobotState()`. 
- There are some methods `waitForCurrentState()` in `CurrentStateMonitor`, which actually should be renamed to `waitForCompleteState()` as they don't look at recent timestamps.
- I'm still arguing for a true method `syncSceneUpdates()` which incorporates all pending scene updates in the callback queue. I agree to @v4hn that this doesn't guarantee that all scene updates up to a given timestamp are incorporated, because - due to network delays - ROS messages might be received delayed. The only guarantee for a synchronous scene update is the new method `applyPlanningScene()`.
- I added some integration tests in moveit_ros/test. To this end, I imported a Fanuc robot description and moveit_config from ROS industrial. Please check, if that's fine.

rhaschke committed:
moved validatePlan()

from MoveGroupInterface (client side)
to MoveGroup's ExecutionServiceCapability (server side)

This checks validity of plan no matter which client called.
Placing it in MoveGroupContext will allow re-use by other capabilities.

rhaschke committed:
reworked syncSceneUpdates() to monitor the robot state

Primary objective of syncSceneUpdates() is to receive a recent robot state.
If a current state monitor is active, this is all to monitor.
If not, scene updates are monitored, in particular the timestamp of their RobotState member.

rhaschke committed:
remove syncSceneUpdates() after trajectory execution

partially reverts f38a55786319a33fb2157e77cd1c49b58daaba8f

rhaschke committed:
add debugging output

rhaschke committed:
reworked syncSceneUpdates()

- new CurrentStateMonitor::waitForCurrentState(ros::Time)
- simply wait for state update to reach scene
- fixed validatePlan(): use new CSM::waitForCurrentState()

rhaschke committed:
fixed unnecessary waiting

If the robot doesn't move, an update in CSM is not forwarded to the planning scene.
Hence, we always wait until timeout for a recent last_robot_motion_time_.
If we ensure that an update, if it is triggered by CSM, is directly
passed to the scene, we can early return true.

This partially reverts commit 907980a392d352f5e2ebb6a0b6906c85d9c7d72c.

rhaschke committed:
MoveGroupInterface::getCurrentState(): wait for current state

davetcoleman commented:
I think the integration tests make this PR way too big - should be separate as discussion on call

rhaschke committed:
renamed syncSceneUpdates() to waitForCurrentRobotState()

new syncSceneUpdates() indeed also waits for all pending scene updates to be processed

v4hn commented on move_group/include/moveit/move_group/move_group_context.h:
```diff
@@ -38,6 +38,7 @@
 #define MOVEIT_MOVE_GROUP_CONTEXT_
 
 #include <moveit/macros/class_forward.h>
+#include <trajectory_msgs/JointTrajectory.h>
```
please add the new dependency to the package.xml

v4hn commented on move_group/src/move_group_context.cpp:
```diff
@@ -89,3 +89,41 @@ bool move_group::MoveGroupContext::status() const
+  {
+    ROS_WARN("Failed to receive full current joint state");
+    return false;
+  }
```
Shouldn't we check `csm->getCurrentStateTime()` and compare to `ros::Time::now()` here to see if our state is up to date?

v4hn commented on move_group/src/move_group_context.cpp:
```diff
@@ -89,3 +89,41 @@ bool move_group::MoveGroupContext::status() const
+  robot_state::RobotStatePtr current_state;
+  if (!csm->waitForCurrentState(1.0) || !(current_state = csm->getCurrentState()))
+  {
+    ROS_WARN("Failed to receive full current joint state");
```
Please make that a `ROS_ERROR` (the failure cases below are).
Also please add something like "Failed to verify the proposed trajectory is valid" to the log output.
This should make the problem more apparent.

v4hn commented on move_group/src/move_group_context.cpp:
```diff
@@ -89,3 +89,41 @@ bool move_group::MoveGroupContext::status() const
+      return false;
+    }
+    // TODO: check multi-DoF joints
+    if (fabs(current_state->getJointPositions(jm)[0] - positions[i]) > std::numeric_limits<float>::epsilon())
```
I'm not entirely sure here, but this is a _huge_ requirement.
Is _epsilon_ the right thing to compare to?
I would prefer a rosparam with a default value of something like `1e-3` here.
If our hardware driver send us an update between planning and execution it is actually quite likely we will not see the _exact_ same joint states here, right?

davetcoleman replied:
+1

v4hn commented on move_group/src/move_group_context.cpp:
```diff
@@ -89,3 +89,41 @@ bool move_group::MoveGroupContext::status() const
+    // TODO: check multi-DoF joints
+    if (fabs(current_state->getJointPositions(jm)[0] - positions[i]) > std::numeric_limits<float>::epsilon())
+    {
+      ROS_ERROR("Trajectory start deviates from current robot state");
```
Could you add "Invalid Trajectory" to the beginning of both error messages?

v4hn commented on planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h:
```diff
@@ -336,8 +336,13 @@ class PlanningSceneMonitor : private boost::noncopyable
+   * Hence, you can specify a timeout to wait for those updates. Default is 1s.
+   */
+  bool syncSceneUpdates(const ros::Time &t = ros::Time::now(),
+                        const ros::WallDuration &timeout = ros::WallDuration(1.));
```
I would prefer not to have the default value for the timeout here.
Leaving it to the user makes it more apparent that there actually _is_ a timeout value involved.
No hard feelings here though.

v4hn commented on move_group/src/move_group_context.cpp:
```diff
@@ -89,3 +89,41 @@ bool move_group::MoveGroupContext::status() const
+  robot_state::RobotStatePtr current_state;
+  if (!csm->waitForCurrentState(ros::Time::now()) || !(current_state = csm->getCurrentState()))
+  {
+    ROS_WARN("Failed to receive full current joint state");
```
Please make that a `ROS_ERROR` (the failure cases below are).
Also please add something like "Failed to verify the proposed trajectory is valid" to the log output.
This should make the problem more apparent.

v4hn commented on planning/planning_scene_monitor/include/moveit/planning_scene_monitor/current_state_monitor.h:
```diff
@@ -136,7 +137,11 @@ class CurrentStateMonitor
+
+  // TODO: rename functions to waitForCompleteState
+  /** @brief Wait for at most \e wait_time seconds until the complete robot state is known. Return true if the full state is known */
   bool waitForCurrentState(double wait_time) const;
```
We _could_ mark this function as deprecated in indigo, to point out that it does not do what it says.
We can't remove the function in indigo, but I think it is a good idea to mark it as deprecated and
make it call the renamed function `waitForCompleteState` in indigo.

davetcoleman replied:
+1 to using deprecation marcos to improve readability

v4hn commented on planning/planning_scene_monitor/src/current_state_monitor.cpp:
```diff
@@ -262,8 +262,26 @@ bool planning_scene_monitor::CurrentStateMonitor::haveCompleteState(const ros::D
+  ros::WallDuration elapsed;
+  ros::WallDuration timeout(wait_time);
+
+  boost::mutex::scoped_lock lock(state_update_lock_);
```
you don't need (and actually should use) that lock here, do you?

v4hn replied:
Looking into this a bit I don't see how this function could ever return without timeout.
`jointStateCallback` requires the lock for a successful update of `current_state_time_` et al.
I would propose to use `getCurrentStateTime()` in the condition below and remove the lock.

rhaschke replied:
Waiting on the state_update_condition_ releases the lock and reaquires it if returning. 
This is perfectly fine.

v4hn commented on planning/planning_scene_monitor/src/planning_scene_monitor.cpp:
```diff
@@ -501,6 +502,8 @@ bool planning_scene_monitor::PlanningSceneMonitor::newPlanningSceneMessage(const
 
     last_update_time_ = ros::Time::now();
     last_robot_motion_time_ = scene.robot_state.joint_state.header.stamp;
+    ROS_DEBUG_STREAM_NAMED("PSM", "scene update " << fmod(last_update_time_.toSec(), 10.)
```
Please rename all the stream output to "planning_scene_monitor". This named-log is already used in the psm and I don't see a reason to add two different names for one class.

davetcoleman replied:
+1

rhaschke replied:
Good point.

v4hn commented on planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h:
```diff
@@ -521,6 +532,7 @@ class PlanningSceneMonitor : private boost::noncopyable
   ros::CallbackQueue                    callback_queue_;
   boost::scoped_ptr<ros::AsyncSpinner>  spinner_;
   ros::Time                             last_robot_motion_time_; /// Last time the robot has moved
+  bool                                  enforce_next_state_update_;
```
~~You only ever _assign_ `enforce_next_state_update_` in this request, but never read from it.
This doesn't make sense.~~
Ignore, I'm blind and this makes total sense.

v4hn commented on planning/planning_scene_monitor/src/planning_scene_monitor.cpp:
```diff
@@ -818,35 +821,70 @@ void planning_scene_monitor::PlanningSceneMonitor::currentWorldObjectUpdateCallb
 
-  // Robot state updates in the scene are only triggered by the state monitor on changes of the state.
-  // Hence, last_state_update_time_ might be much older than current_state_monitor_ (when robot didn't moved for a while).
+  // Eventually there is no state monitor. In this case state updates are received as part of scene updates only.
```
"Eventually" doesn't make sense here (false friend for Germans), "There is no state monitor" is enough.

davetcoleman replied:
lol, yea doesn't make sense

v4hn commented on planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h:
```diff
@@ -336,8 +336,19 @@ class PlanningSceneMonitor : private boost::noncopyable
+   * Additionally to waitForCurrentRobotState() this processes also all already received scene updates
+   * as long as wait_time is not exceeded.
+   */
+  bool syncSceneUpdates(const ros::Time &t = ros::Time::now(), double wait_time = 1.);
```
-1 But you knew that before, because we disagreed on that method before. :)
The problem here is that PlanningScenes do not know the concept of "submission time" and there is, even in theory, no way to make sure we updated the scene "to the current state", because packages could still linger about in the network and this function will miss these updates.
Thus,
- the function does not hold what its name promises (I know at least 5 students from my lab who wouldn't care about the "already received" part in the comment)
- users still have to make sure their updates arrived using an external mechanisms (many users prefer `sleep(5.0)`)
- Many users would surely call this function instead of making sure things are alright on their own and during their next demo the Heisenbug strikes and they miss their decisive scene update.

@davetcoleman , @mikeferguson I suppose we need a third opinion on whether we should have that function in the API.

I consider it to be too misleading and broken by design, @rhaschke considers it to be a reasonable addition.

davetcoleman replied:
I'm confused what this function is doing, perhaps due to the wording of the comments. I'll attempt to rewrite:

```
\brief Wait for current robot state and all pending scene updates to be processed. 

In addition to including the waitForCurrentRobotState() functionality, this function also processes all the already received scene updates - until the wait_time is exceeded
```

v4hn replied:
@davetcoleman no, it does **_not**_ sleep, but instead returns early when there is nothing more in the node's queue - that does not mean though, that no message is on its way...

davetcoleman replied:
Ok:

> \brief Wait for current robot state and all pending scene updates to be processed. 
> 
> In addition to including the waitForCurrentRobotState() functionality, this function also processes all the already received scene updates - _unless_ the wait_time is exceeded

davetcoleman replied:
So perhaps the function should be called `waitForAlreadyReceivedUpdates()`?

What is @rhaschke use case for this? This seems mildly useful still.

rhaschke replied:
> > -  bool syncSceneUpdates(const ros::Time &t = ros::Time::now(), double wait_time = 1.);
> 
> So perhaps the function should be called 
> |waitForAlreadyReceivedUpdates()|?
> 
> What is @rhaschke https://github.com/rhaschke use case for this? 
> This seems mildly useful still.

The idea is to call this function before starting to plan in order to 
incorporate all previously send scene updates.

davetcoleman replied:
-1 it doesn't actually solve a problem - like @v4hn said it would still require a sleep with a guessed amount of time because you won't know if your messages were received yet when you call this function

rhaschke replied:
If scene updates and planning requests are triggered from the very same program, the planning request is definitely processed after the update requests, isn't it?.
Thus it makes sense to have that function in place to ensure that all previous updates are considered for planning.

Of course, @v4hn is right, if another source for scene updates is involved. However, in that case the program sending the planning request cannot sync these updates beforehand as well.

v4hn replied:
On Wed, Aug 17, 2016 at 12:02:41PM -0700, Robert Haschke wrote:

> If scene updates and planning requests are triggered from the very same program,
> the planning request is definitely processed after the update requests, isn't it?.

While this might be true (I'm not sure and it might depend on the order in which the subscriber/publisher are created),
I don't think anyone should rely on this.

> if another source for scene updates is involved. However, in that case the program sending the planning request cannot sync these updates beforehand as well.

Sure, he can hold a PSM depending on `move_group`'s output and wait for expected updates to become visible.
Lousy solution, still, I would always prefer synchronous updates...

rhaschke commented:
I'm afraid that I cannot handle the comments before next week. Please be patient.

v4hn commented:
@rhaschke too bad. Sorry it took me so long :(
@130s is it ok, if this will take another week before a new indigo release?
Releasing without this fixed-up request is not an option.

davetcoleman commented:
Hopefully before August 5th...

v4hn commented:
@davetcoleman , @130s, @mikeferguson / whoever feels responsible. This is a bit critical, so please read and comment.

@davetcoleman this is not going to happen before Friday,
I've been told @rhaschke is on vacation _starting_ this Monday.

He will be really pissed because of that, but at the moment I'm considering a revert of the whole list of commits for #442 up to now. Maybe that's what we should have done directly after the unreviewed merge...
We have at least one open bug with respect to these changes by @dornhege, a list of things that should be improved, ABI breakage w.r.t. the last release and I just noticed a segfault in the RViz plugin in my setup when removing the MotionPlanning display, that gdb blames on the new `spinner_` that has been added to the PlanningSceneMonitor. I'm not sure we **can** have that spinner there, because [ros keeps a recursive mutex on spinners](https://github.com/ros/ros_comm/blob/kinetic-devel/clients/roscpp/src/libros/spinner.cpp#L36), and there _must_ only be one thread that keeps spinners. I'm pretty sure this might have broken API in a totally unintuitive way (definitely not acceptable in indigo), because if you keep a AsyncSpinner around from your main thread and create the monitor from a different thread things probably break.

Overall, this is no state in which we should let people work on "minor" issues and the whole "fix the start state" patch is clearly not ready for a release.

I therefore propose to revert these change-sets and try to add only the part that makes the ExecutionController check whether the trajectory it is about to execute starts somewhere around the current state. This should be possible without most of this code and without breakage.
Later on we can focus on getting this to work without this many issues..

davetcoleman commented:
Yes, after seeing https://github.com/ros-planning/moveit_ros/issues/736 and possibly related https://github.com/ros-planning/moveit/issues/10 I have been worried about the state of the PlanningSceneMonitor. Considering it was never fully reviewed I do not think we should feel guilty about the revert. +1

dornhege commented:
+1 for the revert. Ideally before the merge and unless it is shown that #736 is not an issue in general or independent from this, we cannot release.
The PR is still there to be worked on, but this seems to be not trivial to get right.

rhaschke commented:
Looks like [AsyncSpinner isn't really usable](https://github.com/ros/ros_comm/issues/277) as intended. As I can't work on the issue in the next days, go for the revert as suggested by @v4hn.

davetcoleman commented:
@v4hn can you do the revert today before tomorrow's merge?

v4hn commented:
I opened a request to revert the previous merge commits in #742 .

I'll close this request. We will have to look through the changes again anyway and rebase/reformat them for the merged repo.

rhaschke commented:
@v4hn @dornhege @davetcoleman 
I'm back from vacation and I'm going to look into the `AsyncSpinner` issue now.

Obviously any use of an `AsyncSpinner` is currently not safe: If they are started from different threads, only the very first one will actually be started. Hence it depends on the order of their start, which one will be active, which is not acceptable. As there are multiple `AsyncSpinners` used throughout the source, #736 reported by @dornhege might be only the tip of the iceberg.

I will try to fix that issue upstream in `ros_comm` and then analyze, why it failed in #736 but not for me.

v4hn commented:
On Mon, Aug 15, 2016 at 12:32:20AM -0700, Robert Haschke wrote:

> I will try to fix that issue upstream in `ros_comm` and then analyze, why it failed in #736 but not for me.

Thanks! This is clearly a bug in ros_comm and the commit that introduced the global mutex there is at fault.
The trouble is, it might be hard to remove it again without understanding why it was introduced in the first place..
Maybe it's worth a try to contact the author? It has been a couple of years, but maybe he remembers the original bug (if there was one)?

davetcoleman commented:
Welcome back @rhaschke, hope vacation was good! Thanks for continuing to work on this

rhaschke commented:
The other usage of `AsyncSpinners` in `moveit_ros` source tree is uncritical. They mostly belong to stand-alone programs. I think, the reason that @dornhege observed the bug, but not me, was that he was running the whole perception pipeline too. The `perception/mesh_filter` instantiates its own PlanningSceneMonitor and thus has thrown the error.

I prepared https://github.com/ros/ros_comm/pull/867 and hope that they will consider it timely.

v4hn commented:
Wow, given the scope of your patch @rhaschke I'm pretty sure this will not be merged in indigo, not sure about kinetic. But we should probably fix the safety issue in indigo...
